vkubelet/vkubelet.go

```diff
@@ -165,77 +169,77 @@ func (s *Server) registerNode(ctx context.Context) error {
+		FieldSelector: fields.OneTermEqualSelector("spec.nodeName", s.nodeName).String(),
+	}
 
-	sig := make(chan os.Signal, 1)
-	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
-	go func() {
-		<-sig
-		shouldStop = true
-		s.Stop()
-	}()
+	pods, err := s.k8sClient.CoreV1().Pods(s.namespace).List(opts)
+	if err != nil {
+		log.G(ctx).WithError(err).Error("Unable to get all pods")
+		return
+	}
 
-	for {
-		opts := metav1.ListOptions{
-			FieldSelector: fields.OneTermEqualSelector("spec.nodeName", s.nodeName).String(),
-		}
+	s.resourceManager.SetPods(pods)
+	s.reconcile(ctx)
 
-		pods, err := s.k8sClient.CoreV1().Pods(s.namespace).List(opts)
-		if err != nil {
-			return pkgerrors.Wrap(err, "error getting pod list")
-		}
-		s.resourceManager.SetPods(pods)
-		s.reconcile(ctx)
+	opts.ResourceVersion = pods.ResourceVersion
 
-		opts.ResourceVersion = pods.ResourceVersion
-		s.podWatcher, err = s.k8sClient.CoreV1().Pods(s.namespace).Watch(opts)
-		if err != nil {
-			return pkgerrors.Wrap(err, "failed to watch pods")
-		}
+	var controller cache.Controller
+	_, controller = cache.NewInformer(
 
-	loop:
-		for {
-			select {
-			case ev, ok := <-s.podWatcher.ResultChan():
-				if !ok {
-					if shouldStop {
-						log.G(ctx).Info("Pod watcher is stopped")
-						return nil
-					}
+		&cache.ListWatch{
 
-					log.G(ctx).Error("Pod watcher connection is closed unexpectedly")
-					break loop
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				if controller != nil {
+					opts.ResourceVersion = controller.LastSyncResourceVersion()
 				}
 
-				log.G(ctx).WithField("type", ev.Type).Debug("Pod watcher event is received")
-				reconcile := false
-				switch ev.Type {
-				case watch.Added:
-					reconcile = s.resourceManager.UpdatePod(ev.Object.(*corev1.Pod))
-				case watch.Modified:
-					reconcile = s.resourceManager.UpdatePod(ev.Object.(*corev1.Pod))
-				case watch.Deleted:
-					reconcile = s.resourceManager.DeletePod(ev.Object.(*corev1.Pod))
-				}
+				return s.k8sClient.Core().Pods(s.namespace).List(opts)
+			},
 
-				if reconcile {
-					s.reconcile(ctx)
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				if controller != nil {
+					opts.ResourceVersion = controller.LastSyncResourceVersion()
 				}
-			}
-		}
 
-		time.Sleep(5 * time.Second)
-	}
+				return s.k8sClient.Core().Pods(s.namespace).Watch(opts)
+			},
+		},
 
-}
+		&corev1.Pod{},
+
+		1*time.Minute,
+
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) {
+				s.onAddPod(ctx, obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				s.onUpdatePod(ctx, newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				s.onDeletePod(ctx, obj)
+			},
+		},
+	)
 
-// Stop shutsdown the server.
-// It does not shutdown pods assigned to the virtual node.
-func (s *Server) Stop() {
-	if s.podWatcher != nil {
-		s.podWatcher.Stop()
+	stopCh := make(chan struct{})
+
+	sig := make(chan os.Signal, 1)
```

Comment:
See https://github.com/virtual-kubelet/virtual-kubelet/pull/368/files#diff-ff7686b39bf90dc2520886fb874371a4R92 where this is moved up into the CLI (and uses context cancelation instead).

Reply:
OK. Seems like lot of conflicts need to handle. I'll review #368 first.